api/sip/__init__.py

```diff
@@ -84,34 +84,39 @@ def __init__(self, library, integration, analytics=None, client=None, connect=Tr
         super(SIP2AuthenticationProvider, self).__init__(
             library, integration, analytics
         )
+
+        self.server = integration.url
+        self.port = integration.setting(self.PORT).int_value
+        self.login_user_id = integration.username
+        self.login_password = integration.password
+        self.location_code = integration.setting(self.LOCATION_CODE).value
+        self.field_separator = integration.setting(self.FIELD_SEPARATOR).value or '|'
+        self.client = client
+
+    def patron_information(self, username, password):
         try:
-            server = None
-            if client:
-                if callable(client):
-                    client = client()
+            if self.client:
+                sip = self.client
             else:
-                server = integration.url
-                port = integration.setting(self.PORT).int_value
-                login_user_id = integration.username
-                login_password = integration.password
-                location_code = integration.setting(self.LOCATION_CODE).value
-                field_separator = integration.setting(
-                    self.FIELD_SEPARATOR).value or '|'
-                client = SIPClient(
-                    target_server=server, target_port=port,
-                    login_user_id=login_user_id, login_password=login_password,
-                    location_code=location_code, separator=field_separator,
-                    connect=connect
+                sip = self.client(
+                    target_server=self.server, target_port=self.port,
+                    login_user_id=self.login_user_id, login_password=self.login_password,
+                    location_code=self.location_code, separator=self.field_separator
                 )
+            sip.connect()
+            sip.login()
+            info = sip.patron_information(username, password)
+            sip.end_session()
+            sip.disconnect()
+            return info
```

Comment:
I was thinking the connect/login/end_session/disconnect loop would happen _inside_ the call to patron_information, but this also works, since the only 'real' SIP method we ever call is patron_information. We can revisit this if we ever start calling another method.

Reply:
I also momentarily thought that this meant your treatment of `sequence_number` was wrong, but I think you have it right. Every time you call connect(), `sequence_number` is reset. So as long as a SIPClient is only used by one thread at a time, everything will be fine.

That said, I think there's no longer any reason to get rid of `socket_lock`. Putting it back would make it possible for multiple threads to use a SIPClient as part of a single session. Experiment with that and see if you can get it to work, but I'm okay with merging this code as is because the circ manager uses one SIPClient per thread.